src/libvirtApi/rhel-images.js

```diff
@@ -0,0 +1,83 @@
+/*
+ * This file is part of Cockpit.
+ *
+ * Copyright (C) 2022 Red Hat, Inc.
+ *
+ * Cockpit is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation; either version 2.1 of the License, or
+ * (at your option) any later version.
+ *
+ * Cockpit is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Cockpit; If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import * as python from "python.js";
+
+import downloadRhelImageScript from 'raw-loader!../scripts/rhsm/download_file_and_report_progress.py';
+import getAccessTokenScript from 'raw-loader!../scripts/rhsm/get_access_token.py';
+import getRhelImageUrlScript from 'raw-loader!../scripts/rhsm/get_rhel_image_url.py';
+
+import {
+    logDebug,
+} from "../helpers.js";
+
+/*
+ * Provider for Red Hat Subscription Manage API.
+ * See https://access.redhat.com/management/api/rhsm
+ */
+
+export function downloadRhelImage(accessToken, url, fileName, downloadDir, isSystem) {
+    logDebug(`Download rhel image`);
+
+    const args = JSON.stringify({
+        accessToken,
+        url,
+        fileName,
+        downloadDir
+    });
+
+    return python.spawn(downloadRhelImageScript, args, { err: "message", superuser: isSystem && "require", environ: ['LC_ALL=C.UTF-8'] });
+}
+
+export function getAccessToken(offlineToken) {
+    logDebug(`Get access token`);
+
+    /*
+     * Alternatively we can use cockpit.http.post, but I can't make it work
+
+    const http = cockpit.http({
+        address: "https://sso.redhat.com/",
+        port: 443,
+    });
+
+    const headers = {
+        grant_type: "refresh_token",
+        client_id: "rhsm-api",
+        refresh_token: offlineToken,
+    };
+
+    return http.post("/auth/realms/redhat-external/protocol/openid-connect/token", undefined, [headers]);\
+    */
+
```

Comment:
@marusak

Reply:
I think all the issues above are fixed, apart from this one. I can't seem to figure out how to make cockpit.http work with an https. Anyway, here, commented out, is how I tried it. I believe the issue is with incorrect headers of the https request.
Without Content-Type set, POST request fails with: { status: 415, reason: "Unsupported Media Type", message: "Unsupported Media Type", problem: null }
With Content-Type set, POST request fails with: { status: 400, reason: "Bad Request", message: "Bad Request" }

@martinpitt have you encountered any issues with cockpit.http using https before?

Reply:
No, I didn't, and the http"s" part is fine -- otherwise you'd get certificate verification errors and not even get as far as getting a HTTP status. Supposedly it sends some different headers. What I would do is to point this against http://localhost:12345 instead of https://sso.redhat.com/, run `nc -l 12345`, run the request with both methods (urllib.request() and cockpit.http.post()), and compare the raw netcat output of the request. This will show the http version, headers, body etc. in the finest possible detail.

Note that this was mostly a suggestion for simplification. I find it ugly and unwieldy to invoke an external Python script, but as long as that works with just "batteries included" python, I won't veto it. What I do veto is to introduce a python-requests dependency, but you already rewrote everything in terms of urllib (thanks!).